File: examples/galactic_unicorn/lava_lamp.cpp
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "pico/stdlib.h"

#include "libraries/pico_graphics/pico_graphics.hpp"
#include "galactic_unicorn.hpp"
#include "okcolor.hpp"

using namespace pimoroni;

PicoGraphics_PenRGB565 graphics(53, 11, nullptr);
GalacticUnicorn galactic_unicorn;

// HSV Conversion expects float inputs in the range of 0.00-1.00 for each channel
// Outputs are rgb in the range 0-255 for each channel
void from_hsv(float h, float s, float v, uint8_t &r, uint8_t &g, uint8_t &b) {
  float i = floor(h * 6.0f);
  float f = h * 6.0f - i;
  v *= 255.0f;
  uint8_t p = v * (1.0f - s);
  uint8_t q = v * (1.0f - f * s);
  uint8_t t = v * (1.0f - (1.0f - f) * s);

  switch (int(i) % 6) {
    case 0: r = v; g = t; b = p; break;
    case 1: r = q; g = v; b = p; break;
    case 2: r = p; g = v; b = t; break;
    case 3: r = p; g = q; b = v; break;
    case 4: r = t; g = p; b = v; break;
    case 5: r = v; g = p; b = q; break;
  }
}

struct blob_t {
  float x, y;
  float r;
  float dx, dy;
};

constexpr int blob_count = 20;

int main() {

  stdio_init_all();

  galactic_unicorn.init();

  // randomise blob start positions, directions, and size
  std::array<blob_t, blob_count> blobs;
  for(auto &blob : blobs) {
    blob.x = rand() % 11;
    blob.y = rand() % 53;
    blob.r = ((rand() % 40) / 10.0f) + 5.0f;
    blob.dx = ((rand() % 2) / 10.0f) - 0.05f;
    blob.dy = ((rand() % 3) / 10.0f) - 0.1f;
  }

  float hue = 0.0f;

  while(true) {
    // allow user to adjust brightness
    if(galactic_unicorn.is_pressed(galactic_unicorn.SWITCH_BRIGHTNESS_UP)) {
      galactic_unicorn.adjust_brightness(+0.01);
    }
    if(galactic_unicorn.is_pressed(galactic_unicorn.SWITCH_BRIGHTNESS_DOWN)) {
      galactic_unicorn.adjust_brightness(-0.01);
    }
    
    uint start_ms = to_ms_since_boot(get_absolute_time());

    // calculate the influence of each blob on the liquid based
    // on their distance to each pixel. this causes blobs to 
    // "merge" into each other when we use fixed thresholds to
    // determine which colour to draw with
    float liquid[11][53] = {0.0f};
    for(auto &blob : blobs) {
      float r_sq = blob.r * blob.r;
      for(int y = 0; y < 53; y++) {
        for(int x = 0; x < 11; x++) {
          float d_sq = (x - blob.x) * (x - blob.x) + (y - blob.y) * (y - blob.y);
          if(d_sq <= r_sq) {
            // add this blobs influence to this pixel
            liquid[x][y] += 1.0f - (d_sq / r_sq);
          }
        }
      }
    }

    // update the blob positions
    for(auto &blob : blobs) {
      blob.x += blob.dx;
      blob.y += blob.dy;

      // if we hit the edge then bounce!
      if(blob.x < 0.0f || blob.x >= 11.0f) {
        blob.dx *= -1.0f;
      }    
      if(blob.y < 0.0f || blob.y >= 53.0f) {
        blob.dy *= -1.0f;
      }    
    }

    // rotate the hue
    hue += 0.001f;

    // calculate dark, medium, and bright shades for rendering the 
    // lava
    uint8_t dark_r, dark_g, dark_b;
    from_hsv(hue, 1.0f, 0.3f, dark_r, dark_g, dark_b);
    uint8_t mid_r, mid_g, mid_b;
    from_hsv(hue, 1.0f, 0.6f, mid_r, mid_g, mid_b);
    uint8_t bright_r, bright_g, bright_b;
    from_hsv(hue, 1.0f, 1.0f, bright_r, bright_g, bright_b);

    // clear the canvas
    graphics.set_pen(0, 0, 0);
    graphics.clear();

    // render the lava
    for(int y = 0; y < 53; y++) {
      for(int x = 0; x < 11; x++) {
        float v = liquid[x][y];

        // select a colour for this pixel based on how much
        // "blobfluence" there is at this position in the liquid
        if(v >= 1.5f) {
          graphics.set_pen(bright_r, bright_g, bright_b);
          graphics.pixel(Point(y, x));
        }else if(v >= 1.25f) {
          graphics.set_pen(mid_r, mid_g, mid_b);
          graphics.pixel(Point(y, x));
        }else if(v >= 1.0f) {
          graphics.set_pen(dark_r, dark_g, dark_b);
          graphics.pixel(Point(y, x));
        }
      }
    }

    uint end_ms = to_ms_since_boot(get_absolute_time());

    printf("rendering took %dms\n", end_ms - start_ms);

    galactic_unicorn.update(graphics);
  }

  return 0;
}

File: micropython/examples/cosmic_unicorn/lava_lamp.py
import time
import random
import math
from cosmic import CosmicUnicorn
from picographics import PicoGraphics, DISPLAY_COSMIC_UNICORN as DISPLAY

'''
A 70s-tastic, procedural rainbow lava lamp.

You can adjust the brightness with LUX + and -.
'''

cu = CosmicUnicorn()
graphics = PicoGraphics(DISPLAY)

blob_count = 10


class Blob():
    def __init__(self):
        self.x = float(random.randint(0, width - 1))
        self.y = float(random.randint(0, height - 1))
        self.r = (float(random.randint(0, 40)) / 10.0) + 5.0
        self.dx = (float(random.randint(0, 2)) / 10.0) - 0.1
        self.dy = (float(random.randint(0, 2)) / 10.0) - 0.05  # positive bias


@micropython.native  # noqa: F821
def setup_portrait():
    global width, height, liquid, blobs
    width = CosmicUnicorn.HEIGHT
    height = CosmicUnicorn.WIDTH
    liquid = [[0.0 for y in range(height)] for x in range(width)]
    blobs = [Blob() for i in range(blob_count)]


hue = 0.0


@micropython.native  # noqa: F821
def from_hsv(h, s, v):
    i = math.floor(h * 6.0)
    f = h * 6.0 - i
    v *= 255.0
    p = v * (1.0 - s)
    q = v * (1.0 - f * s)
    t = v * (1.0 - (1.0 - f) * s)

    i = int(i) % 6
    if i == 0:
        return graphics.create_pen(int(v), int(t), int(p))
    if i == 1:
        return graphics.create_pen(int(q), int(v), int(p))
    if i == 2:
        return graphics.create_pen(int(p), int(v), int(t))
    if i == 3:
        return graphics.create_pen(int(p), int(q), int(v))
    if i == 4:
        return graphics.create_pen(int(t), int(p), int(v))
    if i == 5:
        return graphics.create_pen(int(v), int(p), int(q))


@micropython.native  # noqa: F821
def update_liquid():
    for y in range(height):
        for x in range(width):
            liquid[x][y] = 0.0

    for blob in blobs:
        r_sq = blob.r * blob.r
        blob_y_range = range(max(math.floor(blob.y - blob.r), 0),
                             min(math.ceil(blob.y + blob.r), height))
        blob_x_range = range(max(math.floor(blob.x - blob.r), 0),
                             min(math.ceil(blob.x + blob.r), width))

        for y in blob_y_range:
            for x in blob_x_range:
                x_diff = x - blob.x
                y_diff = y - blob.y
                d_sq = x_diff * x_diff + y_diff * y_diff
                if d_sq <= r_sq:
                    liquid[x][y] += 1.0 - (d_sq / r_sq)


@micropython.native  # noqa: F821
def move_blobs():
    for blob in blobs:
        blob.x += blob.dx
        blob.y += blob.dy

        if blob.x < 0.0 or blob.x >= float(width):
            blob.dx = 0.0 - blob.dx

        if blob.y < 0.0 or blob.y >= float(height):
            blob.dy = 0.0 - blob.dy


@micropython.native  # noqa: F821
def draw_portrait():
    global hue
    hue += 0.001

    dark = from_hsv(hue, 1.0, 0.3)
    mid = from_hsv(hue, 1.0, 0.6)
    bright = from_hsv(hue, 1.0, 1.0)

    for y in range(height):
        for x in range(width):
            v = liquid[x][y]

            # select a colour for this pixel based on how much
            # "blobfluence" there is at this position in the liquid
            if v >= 1.5:
                graphics.set_pen(bright)
            elif v >= 1.25:
                graphics.set_pen(mid)
            elif v >= 1.0:
                graphics.set_pen(dark)
            else:
                graphics.set_pen(0)
            graphics.pixel(y, x)

    cu.update(graphics)


setup_portrait()

cu.set_brightness(0.5)

while True:

    if cu.is_pressed(CosmicUnicorn.SWITCH_BRIGHTNESS_UP):
        cu.adjust_brightness(+0.01)

    if cu.is_pressed(CosmicUnicorn.SWITCH_BRIGHTNESS_DOWN):
        cu.adjust_brightness(-0.01)

    if cu.is_pressed(CosmicUnicorn.SWITCH_A):
        setup_portrait()

    start = time.ticks_ms()

    update_liquid()
    move_blobs()
    draw_portrait()

    # pause for a moment (important or the USB serial device will fail)
    time.sleep(0.001)

    print("total took: {} ms".format(time.ticks_ms() - start))
